Handle NaN in display devices

D4vidH4mm3r opened the pull request:
As I'm not very familiar with the code base yet, I'll just post this to record which parts I've gotten to so far. In any plot, one should decide between replacing nan with some value or dropping it. Same goes for +-infinity values.

Have so far looked at:

- [ ] `fast_plot`
  - [x] `ndarray` with `ndim==1`
  - [ ] `ndarray` with `ndim>1`
  - [ ] `ndarray` with `ndim>1` and `self.with_x`
  - [ ] `xr.DataArray`
- [ ] `hist1d`
- [ ] `hist2d`
- [x] `image`
  - Already just works (on my machine)
- [ ] `plot`
- [ ] `polar_plot`
- [ ] `sorted`
- [ ] `value`
- [x] `waveform`

Developing against `fix/forward-compat` as that allowed me to install locally for testing, but can (rebase and) switch to point to `master` later.

philsmt committed:
Fix packaging incompatibilities with newer Python stacks

philsmt committed:
Fix compatibility with recent PyQt in display.fast_plot

philsmt committed:
Fix compatibility with recent PyQt in display.image

philsmt committed:
Fix compatibility with recent PyQt in display.hist2d

philsmt committed:
Fix compatibility with recent PyQt in display.plot

philsmt committed:
Fix compatibility with recent PyQt in util.simulate.scalar*

D4vidH4mm3r committed:
Handle NaN in fast_plot and waveform

D4vidH4mm3r commented on src/metro/devices/display/fast_plot.py:
```diff
@@ -1235,8 +1235,11 @@ def dataSet(self, d):
     def dataAdded(self, d):
         self.ch_data = d
 
+        nan_replacement = 0
         if isinstance(self.ch_data, xr.DataArray):
             self.idx_data = self.ch_data.data[self.index]
+            if "replace_nan" in self.ch_data.attrs:
+                nan_replacement = self.ch_data.attrs["replace_nan"]
```
One could also allow specifying and utilizing the various `dropna` / `fillna` etc. available for `DataArray`. They don't handle `inf`, though. Also, dropping is tricky.

D4vidH4mm3r commented on src/metro/devices/display/fast_plot.py:
```diff
@@ -1268,6 +1271,9 @@ def dataAdded(self, d):
         if self.idx_data.shape[1] == 0:
             return
 
+        self.idx_data = np.nan_to_num(self.idx_data, copy=True,
+                                      nan=nan_replacement)
+
```
Note: would like to avoid copy.
Cannot replace in-place as destination is not writable.
Ideally, would only copy if any value needs replacing - but first checking that and then copying doesn't seem clever either.

Also: I considered if we could drop instead of replacing.
However, if `self.idx_data.shape[0] > 1`, then we cannot do a simple 1d mask and `np.compress` on x and ys.
And stacking seems tricky.
Unless we drop by logical OR across `axis=0`.

philsmt replied:
Hmmm, this is likely due to the no-copy mechanism used with `pyzmq`. Generally though, I'd also like to avoid encoding `metropc`-specific details in the display devices.

We could either implement it directly in the rendering code, replacing `NaN` by something else, or move this specific into the `euxfel.metropc` device. Another advantage of the rendering code would be that we can actually render gaps rather than replace it. And one could present `np.inf` by an infinite slope before and after.

D4vidH4mm3r commented on src/metro/devices/display/waveform.py:
```diff
@@ -237,6 +237,9 @@ def dataAdded(self, d):
         except TypeError:
             pass
 
+        if not numpy.isfinite(d):
+            return
+
```
Here it's easy to just drop a given value. The plot was almost working as-is: NaN values would cause an exception and just not get plotted, so there would be gap. The curve would break, though.

philsmt replied:
Similar to `fast_plot`, I wonder if it would be useful to visualize this by a gap?

D4vidH4mm3r committed:
Add noexcept, DRY data coord transformation

This was just part of preparation before adding clipping to draw_line, but I got
significant performance boost from, presumably, the noexcept. We can drop the C++ part
if it's not wanted; I just found references easier to work with (syntactically) than
more pointers.

D4vidH4mm3r committed:
Add line segment - box clipping

D4vidH4mm3r committed:
Uses nan_to_num, but scale y based on finite values only

philsmt commented:
Hi, long time no see!

I finally got around to clean up the sorry state this repository was in and made a consistent main branch again. As a bonus, it's compatible with modern Python, Qt, pyqtgraph and more. I think it would still be great to improve NaN behaviour.

Do you think we could rebase and merge what works already?
